src/collections/maple_tree/mod.rs

```diff
@@ -104,66 +131,307 @@ impl<K, V> SparseLeaf<K, V>
             }
         }
 
-        // If there is no slot, return a failure?
+        return None;
+    }
+
+    pub fn search(&mut self, k: &K) -> Option<&V> {
+        for i in 0..CAPACITY {
+            match &self.key[i] {
+                M::Some(v) => {
+                    if v == k {
+                        match &self.value[i] {
+                            M::Some(v) => {
+                                return Some(v);
+                            }
+                            M::None => {
+                                return None;
+                            }
+                        }
+                    }
+                }
+
+                M::None => {}
+            }
+        }
 
-        unimplemented!();
+        None
     }
 
     pub fn update(&mut self, k: K, v: V) -> Option<V> {
-        unimplemented!();
+        for i in 0..CAPACITY {
+            match &self.key[i] {
+                M::Some(v) => {
+                    if v != &k {
+                        continue;
+                    }
+                }
+
+                M::None => {
+                    continue;
+                }
+            }
+
+            let mut nv = M::Some(v);
+
+            mem::swap(&mut self.value[i], &mut nv);
+
+            match nv {
+                M::Some(v) => return Some(v),
+                M::None => return None,
```

Comment:
I thought it was invalid for a key to have a None associated to it? So this could be panic!() instead (or perhaps in #[cfg(test)] it's panic, it's None in release mode. Thoughts?

Reply:
Hmm, this is a hard one because if it is None doesn't that mean that we can't verify the validity of the rest of the tree since we are assuming that a valid key can't have a None value and so we should panic even in release mode?

Reply:
Remember though, the none value would be represented as K: ..., V: Option<T>, so then it would actually be M<Option<T>>, which means that match nv would be matching on M, which can contain an Option, that is None or Some. So my comment is still "correct" I think unless you have something else in mind.

To be 100% sure, we can easily construct a test case of sparseleaf<usize, Option<usize>> or similar, and test adding in none/some values. :)

Reply:
Follow up. Could be fun to create a test for SparseLeaf<Option<usize>, ()> to see how that goes.